Distribution/Server/Features/Security/State.hs

```diff
@@ -267,13 +267,15 @@ updateSecurityState maxAge newUpdate = do
     when (needUpdate oldUpdate newUpdate) $ do
       setTufUpdate newUpdate
       updateSnapshotAndTimestamp (tufUpdateTime newUpdate)
-    SecurityState {
-      securityStateFiles  = Just SecurityStateFiles {
-                              securitySnapshot  = snapshot,
-                              securityTimestamp = timestamp
-                            }
-    } <- State.get
-    return (timestamp, snapshot)
+    st <- State.get
+    case st of
+      SecurityState {
+        securityStateFiles  = Just SecurityStateFiles {
+                                securitySnapshot  = snapshot,
+                                securityTimestamp = timestamp
+                              }
+      } -> return (timestamp, snapshot)
+      _ -> error "updateSecurityState: unexpected state"
```

Comment:
I'm not sure this is the proper fix; what `Monad` are we in here? can it express failure in a more exact way than by a "pure" `error`?

Reply:
That code runs in `acid-state`'s `Update` monad: http://hackage.haskell.org/package/acid-state-0.15.1/docs/Data-Acid.html#t:Update. It's not an instance of `MonadFail`. There is a legacy `fail` method, of course, but it's semantics are not clear to me. I'm not terribly familiar with that code, which is why I opted for the solution that I knew for sure would behave the same way the original version of the code does. I'd rather not screw up security critical code.

Reply:
So, 

```hs
newtype Update st a = Update { unUpdate :: State st a }
    deriving (Monad, Functor, MonadState st)
```

inherits the `Monad` instance of `StateT _ Identity` which does indeed not have a good `fail` operation to offer... so I guess pure exceptions such as `error` are as good as it gets...